### ISSUE TITLE
Directories excluded by excludeRegex can still give permission errors

### ISSUE THREAD

**Issue by jwiegley** [Backblaze/B2_Command_Line_Tool#364]
I have the following sync command:

```
backblaze-b2 sync --keepDays 722 --replaceNewer \
    --excludeRegex '(.*\.DS_Store)|(.*\.Spotlight-V100.*)|(.*\.Trashes.*)|(.*\.fseventsd.*)|(.*\.DocumentRevisions-V100.*)|(.*\.TemporaryItems)' \
    /Volumes/mybook/Audio b2://XXX
```

However, despite Trashes being ignored, I still get:

```
WARNING: /Volumes/mybook/Audio/.Trashes could not be accessed (no permissions to read?)
```

**Comment by ppolewicz** [comment]
That's a bug. The way `--excludeRegex` is implemented, it first makes a list of everything (including things that sync shouldn't touch) and then it filters items. This can cause a directory with a very large number of files to be listed unnecessarily, wasting time, wasting CPU and potentially causing warnings as in your case.

**Comment by dmanikowski-reef** [comment]
Hi guys, I looked into it. It is quite easy to replace the list @ppolewicz is talking about with the generator but there is a minor issue. List of files to be copied is prepared using all_files() method of the LocalFolder class. As far as I know this method should return SORTED list of all files in the directory which we are unable to achieve using generator. We can sort the list afterwards if needed but I don't know if this would be expected behavior. Is this sorting really needed in all_files() method?

**Comment by ppolewicz** [comment]
I believe sync uses the sorting later on. We can move the scanning to another (sub)method which returns a generator, filter right there and then cast to list and perform the sorting in all_files()

**Comment by bwbeach** [comment]
The doc for `sync` says:

    You can specify --excludeRegex to selectively ignore files that
    match the given pattern. Ignored files will not copy during
    the sync operation. The pattern is a regular expression
    that is tested against the full path of each file.

I think that excluding `.*\.DS_Store` should not exclude `.DS_Store/a.txt`, because the regex doesn't match the full file name.  To exclude it, you would need a regex like `.*\.DS_Store/.*`.

**Comment by bwbeach** [comment]
#407 took care of this one.

**Comment by marcobarriosfi** [comment]
This problem still exists.

**Comment by ppolewicz** [comment]
@marcobarriosfi which version of the CLI are you using?

**Comment by marcobarriosfi** [comment]
Seems to be 1.1.0-1. Installed it via apt.

**Comment by ppolewicz** [comment]
It was fixed in v1.2.0.

If the problem is easily reproducible, please try the newest version before reporting an issue.

**Comment by marcobarriosfi** [comment]
@ppolewicz Apologies. I installed the CLI yesterday via apt so I assumed that I was running latest/relatively up-to-date version. The CLI itself does not show the version number or at least I did not spot it.
